File: src/router.zig
//! Advanced Router module for Wraith
//! Handles HTTP/3 request routing with host/path/header matching

const std = @import("std");
const Allocator = std.mem.Allocator;

pub const Router = struct {
    allocator: Allocator,
    routes: std.ArrayList(Route),
    middleware: std.ArrayList(Middleware),
    
    const Self = @This();

    pub fn init(allocator: Allocator) Self {
        return Self{
            .allocator = allocator,
            .routes = std.ArrayList(Route).init(allocator),
            .middleware = std.ArrayList(Middleware).init(allocator),
        };
    }

    pub fn deinit(self: *Self) void {
        for (self.routes.items) |route| {
            self.allocator.free(route.path);
            if (route.host) |host| {
                self.allocator.free(host);
            }
        }
        self.routes.deinit();
        self.middleware.deinit();
    }

    pub fn addRoute(self: *Self, config: RouteConfig) !void {
        const route = Route{
            .path = try self.allocator.dupe(u8, config.path),
            .host = if (config.host) |h| try self.allocator.dupe(u8, h) else null,
            .method = config.method,
            .handler = config.handler,
            .route_type = config.route_type,
            .priority = config.priority,
            .middleware = config.middleware,
        };
        
        try self.routes.append(route);
        
        // Sort routes by priority (higher priority first)
        std.sort.insertion(Route, self.routes.items, {}, routePriorityLessThan);
    }
    
    fn routePriorityLessThan(_: void, a: Route, b: Route) bool {
        return a.priority > b.priority;
    }

    pub fn match(self: *Self, request: *const RoutingRequest) ?RouteMatch {
        for (self.routes.items) |*route| {
            if (self.routeMatches(route, request)) {
                return RouteMatch{
                    .route = route,
                    .params = self.extractParams(route, request) catch null,
                };
            }
        }
        return null;
    }
    
    fn routeMatches(self: *Self, route: *const Route, request: *const RoutingRequest) bool {
        // Method matching
        if (route.method != .ANY and route.method != request.method) {
            return false;
        }
        
        // Host matching
        if (route.host) |expected_host| {
            if (request.host == null or !std.mem.eql(u8, expected_host, request.host.?)) {
                return false;
            }
        }
        
        // Path matching
        return self.pathMatches(route.path, request.path);
    }
    
    fn pathMatches(self: *Self, pattern: []const u8, path: []const u8) bool {
        _ = self;
        
        // Exact match
        if (std.mem.eql(u8, pattern, path)) {
            return true;
        }
        
        // Wildcard matching
        if (std.mem.endsWith(u8, pattern, "/*")) {
            const prefix = pattern[0..pattern.len - 2];
            return std.mem.startsWith(u8, path, prefix);
        }
        
        // Parameter matching (e.g., /users/:id)
        if (std.mem.indexOf(u8, pattern, ":")) |_| {
            return self.parameterMatches(pattern, path);
        }
        
        return false;
    }
    
    fn parameterMatches(self: *Self, pattern: []const u8, path: []const u8) bool {
        _ = self;
        
        var pattern_parts = std.mem.split(u8, pattern, "/");
        var path_parts = std.mem.split(u8, path, "/");
        
        while (pattern_parts.next()) |pattern_part| {
            const path_part = path_parts.next() orelse return false;
            
            if (pattern_part.len > 0 and pattern_part[0] == ':') {
                // Parameter segment - matches any non-empty path segment
                if (path_part.len == 0) return false;
            } else {
                // Literal segment - must match exactly
                if (!std.mem.eql(u8, pattern_part, path_part)) {
                    return false;
                }
            }
        }
        
        // Ensure no extra path segments
        return path_parts.next() == null;
    }
    
    fn extractParams(self: *Self, route: *const Route, request: *const RoutingRequest) !?std.StringHashMap([]const u8) {
        var params = std.StringHashMap([]const u8).init(self.allocator);
        
        if (std.mem.indexOf(u8, route.path, ":") == null) {
            return null; // No parameters to extract
        }
        
        var pattern_parts = std.mem.split(u8, route.path, "/");
        var path_parts = std.mem.split(u8, request.path, "/");
        
        while (pattern_parts.next()) |pattern_part| {
            const path_part = path_parts.next() orelse break;
            
            if (pattern_part.len > 0 and pattern_part[0] == ':') {
                const param_name = pattern_part[1..];
                try params.put(try self.allocator.dupe(u8, param_name), try self.allocator.dupe(u8, path_part));
            }
        }
        
        return params;
    }
    
    pub fn addMiddleware(self: *Self, middleware: Middleware) !void {
        try self.middleware.append(middleware);
    }
};

pub const HttpMethod = enum {
    GET,
    POST,
    PUT,
    DELETE,
    HEAD,
    OPTIONS,
    PATCH,
    ANY,
};

pub const RouteType = enum {
    static,      // Serve static files
    proxy,       // Reverse proxy to backend
    redirect,    // HTTP redirect
    api,         // API endpoint
    websocket,   // WebSocket upgrade
};

pub const RouteConfig = struct {
    path: []const u8,
    host: ?[]const u8 = null,
    method: HttpMethod = .ANY,
    handler: RouteHandler,
    route_type: RouteType = .api,
    priority: u8 = 50, // 0-255, higher = higher priority
    middleware: []const Middleware = &.{},
};

pub const Route = struct {
    path: []const u8,
    host: ?[]const u8,
    method: HttpMethod,
    handler: RouteHandler,
    route_type: RouteType,
    priority: u8,
    middleware: []const Middleware,
};

pub const RoutingRequest = struct {
    path: []const u8,
    host: ?[]const u8,
    method: HttpMethod,
    headers: std.StringHashMap([]const u8),
    remote_addr: ?[]const u8,
};

pub const RouteMatch = struct {
    route: *const Route,
    params: ?std.StringHashMap([]const u8),
};

pub const RouteHandler = *const fn (request: *const RoutingRequest, params: ?std.StringHashMap([]const u8)) anyerror!RouteResponse;

pub const RouteResponse = struct {
    status: u16,
    body: []const u8,
    headers: std.StringHashMap([]const u8),
    content_type: []const u8 = "text/html",
};

pub const Middleware = struct {
    name: []const u8,
    handler: MiddlewareHandler,
};

pub const MiddlewareHandler = *const fn (request: *RoutingRequest, next: *const fn() anyerror!RouteResponse) anyerror!RouteResponse;

/// Create a basic router with common routes
pub fn createDefaultRouter(allocator: Allocator) !Router {
    var router = Router.init(allocator);
    
    // Add health check endpoint
    try router.addRoute(.{
        .path = "/health",
        .method = .GET,
        .handler = healthCheckHandler,
        .priority = 100,
    });
    
    // Add status endpoint
    try router.addRoute(.{
        .path = "/status",
        .method = .GET,
        .handler = statusHandler,
        .priority = 100,
    });
    
    // Add catch-all for static files
    try router.addRoute(.{
        .path = "/*",
        .method = .GET,
        .handler = staticFileHandler,
        .route_type = .static,
        .priority = 1, // Lowest priority
    });
    
    return router;
}

fn healthCheckHandler(request: *const RoutingRequest, params: ?std.StringHashMap([]const u8)) !RouteResponse {
    _ = request;
    _ = params;
    
    var headers = std.StringHashMap([]const u8).init(std.heap.page_allocator);
    try headers.put("content-type", "application/json");
    
    return RouteResponse{
        .status = 200,
        .body = "{\"status\":\"ok\",\"protocol\":\"HTTP/3\",\"transport\":\"QUIC\"}",
        .headers = headers,
        .content_type = "application/json",
    };
}

fn statusHandler(request: *const RoutingRequest, params: ?std.StringHashMap([]const u8)) !RouteResponse {
    _ = request;
    _ = params;
    
    var headers = std.StringHashMap([]const u8).init(std.heap.page_allocator);
    try headers.put("content-type", "application/json");
    
    const status_json = 
        \\{
        \\  "server": "Wraith HTTP/3 Proxy",
        \\  "version": "0.1.0",
        \\  "protocol": "HTTP/3",
        \\  "transport": "QUIC",
        \\  "tls": "1.3",
        \\  "uptime": "unknown"
        \\}
    ;
    
    return RouteResponse{
        .status = 200,
        .body = status_json,
        .headers = headers,
        .content_type = "application/json",
    };
}

fn staticFileHandler(request: *const RoutingRequest, params: ?std.StringHashMap([]const u8)) !RouteResponse {
    _ = params;
    
    var headers = std.StringHashMap([]const u8).init(std.heap.page_allocator);
    try headers.put("content-type", "text/html");
    
    // Simple static response for now
    const html = 
        \\<!DOCTYPE html>
        \\<html>
        \\<head><title>🔥 Wraith HTTP/3 Server</title></head>
        \\<body>
        \\<h1>🔥 Welcome to Wraith!</h1>
        \\<p>🚀 Modern QUIC/HTTP3 Reverse Proxy</p>
        \\<p>✅ Protocol: HTTP/3 over QUIC</p>
        \\<p>🔒 TLS: 1.3 (post-quantum ready)</p>
        \\<p>⚡ Transport: UDP (no TCP dependency)</p>
        \\<p>🎯 Built with Zig for maximum performance</p>
        \\<p>📍 Requested path: {s}</p>
        \\</body>
        \\</html>
    ;
    
    const body = try std.fmt.allocPrint(std.heap.page_allocator, html, .{request.path});
    
    return RouteResponse{
        .status = 200,
        .body = body,
        .headers = headers,
        .content_type = "text/html",
    };
}

File: src/router.rs
use crate::{config::Config, proxy::ProxyManager, static_server::StaticFileServer};
use anyhow::Result;
use bytes::Bytes;
use http::{Method, Request, Response, StatusCode};
use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::Arc;
use tracing::{debug, info};

#[derive(Clone)]
pub struct Router {
    config: Config,
    routes: Vec<Route>,
    static_server: Arc<StaticFileServer>,
}

#[derive(Clone)]
struct Route {
    path_pattern: String,
    host_pattern: Option<String>,
    methods: Vec<Method>,
    handler: RouteHandler,
    priority: u32,
}

#[derive(Clone)]
enum RouteHandler {
    Proxy { upstream_name: String },
    Static,
    Health,
    Admin,
}

impl Router {
    pub fn new(config: Config) -> Self {
        let mut routes = Vec::new();

        // Add health check route
        routes.push(Route {
            path_pattern: "/health".to_string(),
            host_pattern: None,
            methods: vec![Method::GET],
            handler: RouteHandler::Health,
            priority: 100,
        });

        // Add proxy routes from config
        if config.proxy.enabled {
            for upstream in &config.proxy.upstreams {
                routes.push(Route {
                    path_pattern: "/*".to_string(),
                    host_pattern: None,
                    methods: vec![],
                    handler: RouteHandler::Proxy {
                        upstream_name: upstream.name.clone(),
                    },
                    priority: 50,
                });
            }
        }

        // Add static file route
        if config.static_files.enabled {
            routes.push(Route {
                path_pattern: "/*".to_string(),
                host_pattern: None,
                methods: vec![Method::GET, Method::HEAD],
                handler: RouteHandler::Static,
                priority: 10,
            });
        }

        // Sort routes by priority (higher first)
        routes.sort_by(|a, b| b.priority.cmp(&a.priority));

        let static_server = Arc::new(StaticFileServer::new(config.static_files.clone()));

        Self {
            config,
            routes,
            static_server,
        }
    }

    pub async fn route_request(
        &self,
        req: Request<()>,
        peer_addr: SocketAddr,
        proxy_manager: Arc<ProxyManager>,
    ) -> Result<(Response<()>, Option<Bytes>)> {
        let path = req.uri().path();
        let method = req.method();
        let host = req
            .headers()
            .get("host")
            .and_then(|h| h.to_str().ok())
            .map(|s| s.to_string());

        debug!(
            "Routing request: {} {} from {}",
            method, path, peer_addr
        );

        // Find matching route
        for route in &self.routes {
            if self.route_matches(&route, path, method, host.as_deref()) {
                return self
                    .handle_route(&route, req, peer_addr, proxy_manager)
                    .await;
            }
        }

        // No route found
        Ok((
            Response::builder()
                .status(StatusCode::NOT_FOUND)
                .body(())?,
            Some(Bytes::from("Not Found")),
        ))
    }

    fn route_matches(
        &self,
        route: &Route,
        path: &str,
        method: &Method,
        host: Option<&str>,
    ) -> bool {
        // Check method
        if !route.methods.is_empty() && !route.methods.contains(method) {
            return false;
        }

        // Check host
        if let Some(ref host_pattern) = route.host_pattern {
            if host.is_none() || !self.pattern_matches(host_pattern, host.unwrap()) {
                return false;
            }
        }

        // Check path
        self.pattern_matches(&route.path_pattern, path)
    }

    fn pattern_matches(&self, pattern: &str, path: &str) -> bool {
        if pattern == path {
            return true;
        }

        // Wildcard matching
        if pattern.ends_with("/*") {
            let prefix = &pattern[..pattern.len() - 2];
            return path.starts_with(prefix);
        }

        // Parameter matching (simplified)
        if pattern.contains(':') {
            let pattern_parts: Vec<&str> = pattern.split('/').collect();
            let path_parts: Vec<&str> = path.split('/').collect();

            if pattern_parts.len() != path_parts.len() {
                return false;
            }

            for (pattern_part, path_part) in pattern_parts.iter().zip(path_parts.iter()) {
                if !pattern_part.starts_with(':') && pattern_part != path_part {
                    return false;
                }
            }

            return true;
        }

        false
    }

    async fn handle_route(
        &self,
        route: &Route,
        req: Request<()>,
        peer_addr: SocketAddr,
        proxy_manager: Arc<ProxyManager>,
    ) -> Result<(Response<()>, Option<Bytes>)> {
        match &route.handler {
            RouteHandler::Health => Ok((
                Response::builder()
                    .status(StatusCode::OK)
                    .header("content-type", "application/json")
                    .body(())?,
                Some(Bytes::from(r#"{"status":"healthy"}"#)),
            )),

            RouteHandler::Static => {
                let path = req.uri().path();
                self.static_server.serve_file(path, req.headers()).await
            }

            RouteHandler::Proxy { upstream_name } => {
                proxy_manager
                    .forward_request(req, upstream_name, peer_addr)
                    .await
            }

            RouteHandler::Admin => Ok((
                Response::builder()
                    .status(StatusCode::OK)
                    .body(())?,
                Some(Bytes::from("Admin endpoint")),
            )),
        }
    }

    pub fn add_route(
        &mut self,
        path: String,
        host: Option<String>,
        methods: Vec<Method>,
        upstream: String,
        priority: u32,
    ) {
        self.routes.push(Route {
            path_pattern: path,
            host_pattern: host,
            methods,
            handler: RouteHandler::Proxy {
                upstream_name: upstream,
            },
            priority,
        });

        // Re-sort routes
        self.routes.sort_by(|a, b| b.priority.cmp(&a.priority));
    }

    pub fn remove_route(&mut self, path: &str, host: Option<&str>) {
        self.routes.retain(|route| {
            !(route.path_pattern == path
                && route.host_pattern.as_deref() == host)
        });
    }

    pub fn list_routes(&self) -> Vec<RouteInfo> {
        self.routes
            .iter()
            .map(|route| RouteInfo {
                path: route.path_pattern.clone(),
                host: route.host_pattern.clone(),
                methods: route
                    .methods
                    .iter()
                    .map(|m| m.to_string())
                    .collect(),
                handler_type: match &route.handler {
                    RouteHandler::Proxy { upstream_name } => {
                        format!("proxy:{}", upstream_name)
                    }
                    RouteHandler::Static => "static".to_string(),
                    RouteHandler::Health => "health".to_string(),
                    RouteHandler::Admin => "admin".to_string(),
                },
                priority: route.priority,
            })
            .collect()
    }
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct RouteInfo {
    pub path: String,
    pub host: Option<String>,
    pub methods: Vec<String>,
    pub handler_type: String,
    pub priority: u32,
}